cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs

```diff
@@ -0,0 +1,180 @@
+
+import           Cardano.Chain.Common (Lovelace, unsafeGetLovelace)
+import           Cardano.CLI.Ops (CliError (..))
+import           Cardano.CLI.Shelley.Run.KeyGen (runGenesisKeyGenDelegate, runGenesisKeyGenGenesis)
+import           Cardano.CLI.Shelley.Parsers (GenesisDir (..), OpCertCounterFile (..),
+                    SigningKeyFile (..), VerificationKeyFile (..))
+import           Cardano.Config.Shelley.Genesis
+
+import           Control.Monad.Trans.Except (ExceptT)
+import           Control.Monad.Trans.Except.Extra (firstExceptT, handleIOExceptT, hoistEither)
+
+import qualified Data.Aeson as Aeson
+import           Data.Aeson.Encode.Pretty (encodePretty)
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import           Data.Char (isDigit)
+import qualified Data.List as List
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as Text
+import           Data.Time.Clock (NominalDiffTime, UTCTime, addUTCTime, getCurrentTime)
+
+import           Ouroboros.Consensus.BlockchainTime (SystemStart (..))
+import           Ouroboros.Consensus.Shelley.Protocol (TPraosStandardCrypto)
+
+import qualified Prelude
+
+import           Shelley.Spec.Ledger.Coin (Coin (..))
+import           Shelley.Spec.Ledger.Keys (GenKeyHash, KeyHash)
+import qualified Shelley.Spec.Ledger.Keys as Ledger
+import           Shelley.Spec.Ledger.TxData (Addr (..))
+
+import           System.Directory (createDirectoryIfMissing, listDirectory)
+import           System.FilePath ((</>), takeExtension)
+
+runGenesisCreate :: GenesisDir -> Word -> Maybe SystemStart -> Lovelace -> ExceptT CliError IO ()
+runGenesisCreate (GenesisDir gendir) count mStart amount = do
+  start <- maybe (SystemStart <$> getCurrentTimePlus30) pure mStart
+  template <- readShelleyGenesis (gendir </> "genesis.spec.json")
+
+  utxoAddrs <- forM [ 1 .. count ] $ createRequiredKeys gendir
+  genDlgs <- readGenDelegsMap gendir
+
+  writeShelleyGenesis (gendir </> "genesis.json") (updateTemplate start amount genDlgs utxoAddrs template)
+
+  liftIO $ putTextLn "Genesis file created, but still need to set genesis delegates and genesis distribution."
+
+-- -------------------------------------------------------------------------------------------------
+
+createRequiredKeys :: FilePath -> Word -> ExceptT CliError IO ShelleyAddress
+createRequiredKeys gendir index = do
+  createDelegateKeys (gendir </> "delegate-keys") index
+  createGenesisKeys (gendir </> "genesis-keys") index
+  createUtxoAddresses (gendir </> "utxo-keys") index
+
+createDelegateKeys :: FilePath -> Word -> ExceptT CliError IO ()
+createDelegateKeys dir index = do
+  liftIO $ createDirectoryIfMissing True dir
+  let strIndex = show index
+  runGenesisKeyGenDelegate
+        (VerificationKeyFile $ dir </> "delegate" ++ strIndex ++ ".vkey")
+        (SigningKeyFile $ dir </> "delegate" ++ strIndex ++ ".skey")
+        (OpCertCounterFile $ dir </> "delegate" ++ strIndex ++ ".opcert")
```

Comment:
could this file end with `.counter`? because otherwise it suggests that this is an actual operational certificate.

Reply:
Yes, good catch.

Reply:
Good idea. Done. Thanks!